doc/config-generic-update-rollback/SONiC_Generic_Config_Update_and_Rollback_Design.md

```diff
@@ -0,0 +1,617 @@
+```
+[
+  {
+    "op": "replace",
+    "path": "/DEVICE_NEIGHBOR/Ethernet8/port",
+    "value": "eth1"
+  },
+  {
+    "op": "add",
+    "path": "/DHCP_SERVERS/4.4.4.4,",remove
+    "path": "/DHCP_SERVERS/2.2.2.2"
+  },
+]
+```
+3) Apply patch using `apply-patch` command
+```
+admin@sonic:~$ apply-patch ./dhcp-changes.patch.json
+```
+4) In case of failure, rollback the config using `rollback` command
+```
+admin@sonic:~$ rollback mycheckpoint --verbose # verbose to see generated patch
+```
+This will internally do a diff, and generate patch of the needed changes and apply using `apply-patch`.
+The patch will be:
+```
+[
+  {
+    "op": "replace",
+    "path": "/DEVICE_NEIGHBOR/Ethernet8/port",
+    "value": "eth0"
+  },
+  {
+    "op": "remove",
+    "path": "/DHCP_SERVERS/4.4.4.4,",
+    "value": {}
+  },
+  {
+    "op": "add",
+    "path": "/DHCP_SERVERS/2.2.2.2"
+  },
+]
+```
+
+### 1.2.2 Container
+
+All the introduced commands will be part of the *python-sonic-utilities* package installed in Debian host O/S.
+
+# 2 Functionality
+
+## 2.1 Target Deployment Use Cases
+
+The `apply-patch` method should help with automating partial config updates, as external systems can generate the update patch, and apply.
+
+The `checkpoint` and `rollback` commands should help improve recoverability, can also be used by external systems to help revert failures during `apply-patch` operation.
+
+Human operators can also leverage the `checkpoint` and `rollback` functionalities while doing updates through the CLI using [SONiC CLI commands](https://github.com/Azure/sonic-utilities/blob/master/doc/Command-Reference.md).
+
+## 2.2 Functional Description
+
+### 2.2.1 Apply-Patch
+The SONiC `apply-patch` command can broadly classified into the following steps
+
+#### Stage-1 Patch Ordering
+As per the JSON standard, the elements described in JSON format are unordered. However, in SONiC there are some tables which can not be processed before another table is updated. For example changes to the VLAN_MEMBER table needs to be processed before the PORTCHANNEL table is processed. If the ordering is not established, the Port Channel delete operation will fail with an error message complaining that the port-channel is still part of a VLAN. So it is essential that the dependency between the tables is identified and the patch is processed in that order.
+
+The `apply-patch` command uses an ordering table which specifies the list of dependent tables that need to be processed before a table is processed. Below is an example depiction of two ConfigDB tables.
+
+```
+{
+  "VLAN" : ["VLAN_MEMBER", "VLAN_INTERFACE"],
+  "PORTCHANNEL" : ["VLAN_MEMBERSHIP", "PORTCHANNEL_MEMBER"]
+}
+```
+A dependency graph is created for all the tables which are in the JSON patch file. The dependency graph is then resolved to find the order in which the updates to tables in the JSON patch are performed. The JSON patch file is then sorted in the resolved order.
```

Comment:
>The dependency graph is then resolved to find the order in which the updates to tables in the JSON patch are performed [](start = 83, length = 118)

This is the interesting part. Will you give details and example?

Reply:
Good idea!

Reply:
Discussed offline, the ordering of updating the tables might be different in case of deletion vs addition

![image](https://user-images.githubusercontent.com/3927651/105886292-158ebb80-5fbf-11eb-854c-88c270230ffb.png)

Reply:
Idea `i` pseudo code to process all lines to be deleted, then processing all lines to be added:
```python
apply-patch(patch):
  current_running_config = get_running_config()
  simulated_config_after_patch = simulate_apply_patch(current_running_config, patch)
  
  common_config = get_common_config(current_running_config, simulated_config_after_patch)

  # change running config to common config, this should be deletion only
  config_replace(common_config, TABLES_DELETION_ORDER)

  # once config is common config, change it to simulated_config_after_patch, this should be addition only
  config_replace(simulated_config_after_patch, TABLES_ADDITION_ORDER)
```

Reply:
I think the pseudo code will be better if just take config_replace to its own method

```python
apply_patch(patch):
  current_running_config = get_running_config()
  simulated_config_after_patch = simulate_apply_patch(current_running_config, patch)
  config_replace(simulated_config_after_patch, current_running_config )

config_replace(target_config, current_config):
  common_config = get_common_config(current_config, target_config)

  # change running config to common config, this should be deletion only
  config_replace(common_config, TABLES_DELETION_ORDER)

  # once config is common config, change it to simulated_config_after_patch, this should be addition only
  config_replace(current_config, TABLES_ADDITION_ORDER)
```

Reply:
The common idea will not work, as it will lead to bunch of deletions followed by bunch of additions. But I might encounter a case where a config cannot be deleted, and can only be replaced. So in the time of deletion/addition of this command the switch might be in a bad state.

I will instead rely on [YANG SONiC Models](https://github.com/Azure/sonic-buildimage/tree/master/src/sonic-yang-models) to fetch the dependency between the configs getting updated in the patch, and decide how they can be updated. For example:
Assume the patch is the following
```
[
 { "op": "remove", "path": "/PORT/Ethernet1" }
 { "op": "remove", "path": "/VLAN_MEMBER/VLAN1000/Ethernet1" },
 { "op": "remove", "path": "/ACL_TABLE/ssh-acl/Ports/Ethernet1" },
]
```

Checking the `YANG` models, we will find that VLAN_MEMBER and ACL are both referencing PORT. So cleaning up of the VLAN_MEMBER and ACL

The new order would be
```
[
 { "op": "remove", "path": "/ACL_TABLE/ssh-acl/Ports/Ethernet1" },
 { "op": "remove", "path": "/PORT/Ethernet1" }
 { "op": "remove", "path": "/VLAN_MEMBER/VLAN1000/Ethernet1" },
]
```

Reply:
updated to a different approach described in the PR